lib/cn/kcompact.c

```diff
@@ -175,176 +79,212 @@ kcompact(struct cn_compaction_work *w, struct kvset_builder *bldr)
-             * currently processing SPAM (which must be be
-             * discarded), and will get FRIES on the next
-             * iteration.
-             *
-             * The following assertions verify that we aren't here
-             * for other reasons (e.g., input kvsets that violate
-             * assumptions about sequence numbers).
-             */
-            assert(dbg_prev_src < curr.src);
-            assert(seq == emitted_seq);
-            if (seq > emitted_seq)
-                seqno_errcnt++;
-
-            assert(vdata != HSE_CORE_TOMB_PFX);
         }
-    }
 
-    prev_kobj = curr.kobj;
+        prev_kobj = curr->kobj;
 
-    dbg_dup = false;
-    dbg_nvals_this_key = 0;
-    dbg_prev_src = curr.src;
+        dbg_dup = false;
+        dbg_nvals_this_key = 0;
+        dbg_prev_idx = idx;
 
-    more = get_next_item(bh, w->cw_inputv, &curr, &w->cw_stats, &err);
-    if (ev(err))
-        goto done;
+        /* Discard the result from pop(). This bin heap is backed by a kv
+         * iterator which has a buffer (kvi_kv) which curr points to. pop()
+         * will not give us the data the we expect because of the backing
+         * buffer, so we call it in order to force all the side effects to
+         * occur, but only grab the next value after pop() calls heapify().
+         */
+        bin_heap2_pop(bh, (void **)&curr);
+        more = bin_heap2_peek(bh, (void **)&curr);
+        if (more) {
+            iter = kvset_cursor_es_h2r(curr->src);
+            for (idx = 0; idx < w->cw_kvset_cnt; idx++) {
+                if (w->cw_inputv[idx] == iter)
+                    break;
+            }
+            assert(idx < w->cw_kvset_cnt);
 
-    if (more) {
-        if (0 == key_obj_cmp(&curr.kobj, &prev_kobj)) {
-            dbg_dup = true;
-            assert(dbg_prev_src <= curr.src);
-            goto get_values;
-        } else if (pt_set && key_obj_cmp_prefix(&pt_kobj, &curr.kobj) != 0) {
-            pt_set = false;
+            if (key_obj_cmp(&curr->kobj, &prev_kobj) == 0) {
+                dbg_dup = true;
+                assert(dbg_prev_idx <= idx);
+                goto values;
+            }
+
+            if (pt_set && key_obj_cmp_prefix(&pt_kobj, &curr->kobj) != 0)
+                pt_set = false;
         }
-    }
 
-    if (emitted_val) {
-        err = kvset_builder_add_key(bldr, &prev_kobj);
-        if (ev(err))
-            goto done;
-        w->cw_stats.ms_keys_out++;
-        w->cw_stats.ms_key_bytes_out += key_obj_len(&prev_kobj);
+        if (emitted_val) {
+            err = kvset_builder_add_key(bldr, &prev_kobj);
+            if (ev(err))
+                goto done;
+            w->cw_stats.ms_keys_out++;
+            w->cw_stats.ms_key_bytes_out += key_obj_len(&prev_kobj);
+        }
     }
 
-    if (more)
-        goto new_key;
-
 done:
     w->cw_vbmap.vbm_waste = w->cw_vbmap.vbm_tot - w->cw_vbmap.vbm_used;
-    bin_heap_destroy(bh);
+    free(sources);
+    bin_heap2_destroy(bh);
```

Comment:
Maybe free the sources after destroying the bin heap?  In case it might touch them for some reason...

Reply:
I used this order because it is the reverse of how the memory was allocated, but I can switch it up

Reply:
Understood, but since we attach the sources to the bin-heap I think it best practice to destroy the bin-heap first.  We have the same similar code elsewhere IIRC...  Might save someone a lot of grief someday :-)